include/picongpu/particles/particleToGrid/combinedAttributes/AverageAttribute.hpp

```diff
@@ -0,0 +1,78 @@
+/* Copyright 2021 Pawel Ordyna
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "picongpu/simulation_defines.hpp"
+
+#include "picongpu/particles/particleToGrid/combinedAttributes/AverageAttribute.def"
+
+#include <limits>
+#include <string>
+#include <vector>
+namespace picongpu
+{
+    namespace particles
+    {
+        namespace particleToGrid
+        {
+            namespace combinedAttributes
+            {
+                template<typename T_Acc>
+                HDINLINE void AverageDivideOperation::operator()(T_Acc const& acc, float1_X& dst, const float1_X& dens)
+                    const
+                {
+                    // avoid dividing by zero. Return zero if density is close to zero.
+                    if(dens[0] <= std::numeric_limits<float_X>::min())
```

Comment:
How do you avoid close to infinite results caused by division of barely zero (e.g.  if you are in a vacuum boarder where macro-particles in neighboring cells just contribute very little due to their shape and thus you divide by a ver y small number?)

Reply:
@PrometheusPi  Well, i hope this won't happen. At least, I didn't see it happen. Whenever density is very small the total quantity should also be very small (it scales with weighting as well) so they should cancel out. 

Btw. it probably should be:
```suggestion
                    if(dens * particles::TYPICAL_NUM_PARTICLES_PER_MACROPARTICLE * CELL_VOLUME 
                        <= std::numeric_limits<float_X>::min())
```

Maybe the limit should also be a bit higher? One could even go with `< 1.0` and set to zero for all cells with the number of real particles below 1 and just properly document this behavior. But I guess that would be some lost information.

Reply:
I think you are right - it should cancel out as long as the power of numerator is not `< 1`. 

I like the `< 1.0` limit, but I agree with you that this might become some lost information and additionally might cause some strange errors for simulations with weightings below 1 (2D TNSA simulations). 

I would think keeping it for now and seeing, whether issues arise at vacuum boarders, might be a more practical approach. I assume you will use this methods extensively in the future for simulations where there is also vacuum involved, correct? Then you will  see whether issues arise and can fix it in a followup pull request.